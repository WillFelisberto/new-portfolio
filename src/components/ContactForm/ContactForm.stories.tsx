import { Meta, StoryObj } from '@storybook/react';
import ContactForm from '.';

const meta: Meta<typeof ContactForm> = {
  title: 'Components/ContactForm',
  component: ContactForm,
  parameters: {
    layout: 'fullscreen',
  },
  args: {
    title: 'Contact Us',
    subtitle: 'We would love to hear from you',
    button: 'Send Message',
    labelName: 'Your Name',
    LabelMensagem: 'Your Message',
    successMessage: 'Message sent successfully!',
    errorMessage: 'Failed to send the message. Please try again later.',
  },
};

export default meta;

type Story = StoryObj<typeof ContactForm>;

export const Default: Story = {};
