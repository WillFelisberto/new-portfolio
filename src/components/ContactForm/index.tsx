'use client';

import { z } from 'zod';
import { useForm } from 'react-hook-form';
import { zodResolver } from '@hookform/resolvers/zod';
import { HiArrowNarrowRight } from 'react-icons/hi';
import { motion } from 'framer-motion';
import axios from 'axios';
import { toast } from 'react-hot-toast';
import SectionTitle from '../SectionTitle';
import Button from '../Button';
import { fadeUpAnimation } from '@/lib/animations';

const contactFormSchema = z.object({
  name: z.string().min(3).max(100),
  email: z.string().email(),
  message: z.string().min(1).max(500),
});

type ContactFormData = z.infer<typeof contactFormSchema>;

type ContactFormType = {
  title: string;
  subtitle: string;
  button: string;
  labelName: string;
  LabelMensagem: string;
  successMessage: string;
  errorMessage: string;
};

const ContactForm = ({
  title,
  subtitle,
  button,
  labelName,
  LabelMensagem,
  successMessage,
  errorMessage,
}: ContactFormType) => {
  const {
    handleSubmit,
    register,
    reset,
    formState: { isSubmitting },
  } = useForm<ContactFormData>({
    resolver: zodResolver(contactFormSchema),
  });

  const onSubmit = async (data: ContactFormData) => {
    try {
      await axios.post('/api/contact', data);
      toast.success(successMessage);
      reset();
    } catch {
      toast.error(errorMessage);
    }
  };

  return (
    <section
      className="flex items-center justify-center bg-gray-950 px-6 py-16 md:py-32"
      id="contact"
    >
      <div className="mx-auto w-full max-w-[420px]">
        <SectionTitle
          subtitle={subtitle}
          title={title}
          className="items-center text-center"
        />
        <motion.form
          className="mt-12 flex w-full flex-col gap-4"
          onSubmit={handleSubmit(onSubmit)}
          {...fadeUpAnimation}
        >
          <input
            placeholder={labelName}
            className="h-14 w-full rounded-lg bg-gray-800 p-4 text-gray-50 ring-sky-600 placeholder:text-gray-400 focus:outline-none focus:ring-2"
            {...register('name')}
          />
          <input
            placeholder="E-mail"
            type="email"
            className="h-14 w-full rounded-lg bg-gray-800 p-4 text-gray-50 ring-sky-600 placeholder:text-gray-400 focus:outline-none focus:ring-2"
            {...register('email')}
          />
          <textarea
            placeholder={LabelMensagem}
            className="h-[138px] w-full resize-none rounded-lg bg-gray-800 p-4 text-gray-50 ring-sky-600 placeholder:text-gray-400 focus:outline-none focus:ring-2"
            {...register('message')}
            maxLength={500}
          />

          <div className="relative mx-auto mt-6 w-max">
            <Button className="relative z-[2]" disabled={isSubmitting}>
              {button}
              <HiArrowNarrowRight size={18} />
            </Button>
            <div className="absolute inset-0 bg-sky-600 opacity-20 blur-2xl" />
          </div>
        </motion.form>
      </div>
    </section>
  );
};
export default ContactForm;
