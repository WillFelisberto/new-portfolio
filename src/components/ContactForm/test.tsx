import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import ContactForm from '.';
import axios from 'axios';
import { toast } from 'react-hot-toast';

jest.mock('axios');
jest.mock('react-hot-toast', () => ({
  toast: {
    success: jest.fn(),
    error: jest.fn(),
  },
}));

jest.mock('../SectionTitle', () => {
  const SectionTitleMock = ({
    title,
    subtitle,
  }: {
    title: string;
    subtitle: string;
  }) => (
    <div>
      <h1>{title}</h1>
      <h2>{subtitle}</h2>
    </div>
  );

  SectionTitleMock.displayName = 'SectionTitleMock';

  return SectionTitleMock;
});

describe('<ContactForm />', () => {
  const mockArgs = {
    title: 'Contact Us',
    subtitle: 'We would love to hear from you',
    button: 'Send Message',
    labelName: 'Your Name',
    LabelMensagem: 'Your Message',
    successMessage: 'Message sent successfully!',
    errorMessage: 'Failed to send the message. Please try again later.',
  };

  it('renders the form fields and button after animation', async () => {
    const { container } = render(<ContactForm {...mockArgs} />);

    // Aguarda a animação ser aplicada e o formulário aparecer
    await waitFor(() => {
      expect(
        screen.getByPlaceholderText(mockArgs.labelName),
      ).toBeInTheDocument();
      expect(screen.getByPlaceholderText('E-mail')).toBeInTheDocument();
      expect(
        screen.getByPlaceholderText(mockArgs.LabelMensagem),
      ).toBeInTheDocument();
      expect(screen.getByText(mockArgs.button)).toBeInTheDocument();
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('submits the form successfully and shows a success message', async () => {
    (axios.post as jest.Mock).mockResolvedValueOnce({});

    render(<ContactForm {...mockArgs} />);

    fireEvent.change(screen.getByPlaceholderText(mockArgs.labelName), {
      target: { value: 'John Doe' },
    });
    fireEvent.change(screen.getByPlaceholderText('E-mail'), {
      target: { value: 'john.doe@example.com' },
    });
    fireEvent.change(screen.getByPlaceholderText(mockArgs.LabelMensagem), {
      target: { value: 'Hello, this is a test message!' },
    });

    fireEvent.click(screen.getByText(mockArgs.button));

    await waitFor(() => {
      expect(axios.post).toHaveBeenCalledWith('/api/contact', {
        name: 'John Doe',
        email: 'john.doe@example.com',
        message: 'Hello, this is a test message!',
      });
      expect(toast.success).toHaveBeenCalledWith(mockArgs.successMessage);
    });
  });

  it('shows an error message when the form submission fails', async () => {
    (axios.post as jest.Mock).mockRejectedValueOnce(
      new Error('Failed to send'),
    );

    render(<ContactForm {...mockArgs} />);

    fireEvent.change(screen.getByPlaceholderText(mockArgs.labelName), {
      target: { value: 'John Doe' },
    });
    fireEvent.change(screen.getByPlaceholderText('E-mail'), {
      target: { value: 'john.doe@example.com' },
    });
    fireEvent.change(screen.getByPlaceholderText(mockArgs.LabelMensagem), {
      target: { value: 'Hello, this is a test message!' },
    });

    fireEvent.click(screen.getByText(mockArgs.button));

    await waitFor(() => {
      expect(axios.post).toHaveBeenCalledWith('/api/contact', {
        name: 'John Doe',
        email: 'john.doe@example.com',
        message: 'Hello, this is a test message!',
      });
      expect(toast.error).toHaveBeenCalledWith(mockArgs.errorMessage);
    });
  });
});
