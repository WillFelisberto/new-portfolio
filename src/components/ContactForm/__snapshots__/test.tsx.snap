// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContactForm /> renders the form fields and button after animation 1`] = `
<section
  class="flex items-center justify-center bg-gray-950 px-6 py-16 md:py-32"
  id="contact"
>
  <div
    class="mx-auto w-full max-w-[420px]"
  >
    <div>
      <h1>
        Contact Us
      </h1>
      <h2>
        We would love to hear from you
      </h2>
    </div>
    <form
      class="mt-12 flex w-full flex-col gap-4"
      style="opacity: 0; transform: translateY(50px);"
    >
      <input
        class="h-14 w-full rounded-lg bg-gray-800 p-4 text-gray-50 ring-sky-600 placeholder:text-gray-400 focus:outline-none focus:ring-2"
        name="name"
        placeholder="Your Name"
      />
      <input
        class="h-14 w-full rounded-lg bg-gray-800 p-4 text-gray-50 ring-sky-600 placeholder:text-gray-400 focus:outline-none focus:ring-2"
        name="email"
        placeholder="E-mail"
        type="email"
      />
      <textarea
        class="h-[138px] w-full resize-none rounded-lg bg-gray-800 p-4 text-gray-50 ring-sky-600 placeholder:text-gray-400 focus:outline-none focus:ring-2"
        maxlength="500"
        name="message"
        placeholder="Your Message"
      />
      <div
        class="relative mx-auto mt-6 w-max"
      >
        <button
          class="flex items-center justify-center gap-2 rounded-lg bg-sky-600 px-4 py-3 text-gray-50 transition-all hover:bg-sky-500 disabled:opacity-50 relative z-[2]"
        >
          Send Message
          <svg
            aria-hidden="true"
            fill="currentColor"
            height="18"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 20 20"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M12.293 5.293a1 1 0 011.414 0l4 4a1 1 0 010 1.414l-4 4a1 1 0 01-1.414-1.414L14.586 11H3a1 1 0 110-2h11.586l-2.293-2.293a1 1 0 010-1.414z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
        <div
          class="absolute inset-0 bg-sky-600 opacity-20 blur-2xl"
        />
      </div>
    </form>
  </div>
</section>
`;
